File: 269.cpp
// There is a new alien language which uses the latin alphabet. However, the
// order among letters are unknown to you. You receive a list of non-empty words
// from the dictionary, where words are sorted lexicographically by the rules of
// this new language. Derive the order of letters in this language.
//
// Example 1:
//
// Input:
// [
//   "wrt",
//   "wrf",
//   "er",
//   "ett",
//   "rftt"
// ]
//
// Output: "wertf"
// Example 2:
//
// Input:
// [
//   "z",
//   "x"
// ]
//
// Output: "zx"
// Example 3:
//
// Input:
// [
//   "z",
//   "x",
//   "z"
// ]
//
// Output: ""
//
// Explanation: The order is invalid, so return "".
// Note:
//
// You may assume all letters are in lowercase.
// You may assume that if a is a prefix of b, then a must appear before b in the
// given dictionary. If the order is invalid, return an empty string. There may
// be multiple valid order of letters, return any one of them is fine.

class Solution {
public:
    string alienOrder(vector<string>& words) {
        unordered_map<char, unordered_set<char>> suc, prev;
        unordered_set<char> all;

        string s;
        for (auto w : words) {
            all.insert(w.begin(), w.end());
            for (int i = 0; i < min(s.size(), w.size()); ++i) {
                if (s[i] != w[i]) {
                    suc[s[i]].insert(w[i]);
                    prev[w[i]].insert(s[i]);
                    break;
                }
            }
            s = w;
        }

        int sum = all.size();
        for (auto p : prev) {
            all.erase(p.first);
        }

        string res;
        while (all.size()) {
            int x = *all.begin();
            all.erase(x);
            res += x;
            for (auto next : suc[x]) {
                prev[next].erase(x);
                if (prev[next].empty()) {
                    all.insert(next);
                }
            }
        }

        return sum == res.size() ? res : "";
    }
};

File: Python3/269.py
class Solution:
    def alienOrder(self, words: List[str]) -> str:
        suc = defaultdict(set)
        prev = defaultdict(set)
        all = set()
        s = ""
        for t in words:
            if len(s) > len(t) and s.startswith(t):
                return ""
            for i in range(min(len(s), len(t))):
                if s[i] != t[i]:
                    suc[s[i]].add(t[i])
                    prev[t[i]].add(s[i])
                    break
            for c in t:
                all.add(c)
            s = t
        cnt = len(all)
        for c in prev.keys():
            all.discard(c)

        res = ""
        while all:
            c = all.pop()
            res += c

            for next in suc[c]:
                prev[next].discard(c)
                if len(prev[next]) == 0:
                    all.add(next)
        return res if len(res) == cnt else ""

